### PR TITLE
feat(bigone): contract implicit api endpoints, fetchMarkets

### PR DESCRIPTION
**Pull request by samgermain**
```
2023-10-27T04:42:03.091Z
Node.js: v18.15.0
CCXT v4.1.29
bigone.fetchMarkets ()
2023-10-27T04:42:12.750Z iteration 0 passed in 3321 ms
            id |                                 uuid |            symbol |         base | quote | settle |    baseId | quoteId | settleId |   type |  spot | margin |  swap | future | option | active | contract | linear | inverse | contractSize | expiry | expiryDatetime | strike | optionType |                            precision |                                                                        limits | created
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
       MOB-BTC | 007b24e6-1e27-4a2f-b4a0-247b4b1f39ee |           MOB/BTC |          MOB |   BTC |        |       MOB |     BTC |          |   spot |  true |  false | false |  false |  false |   true |    false |        |         |              |        |                |        |            |     {"amount":0.01,"price":0.000001} |         {"leverage":{},"amount":{},"price":{},"cost":{"min":0.0001,"max":35}} |        
...
       FILUSDT |                                      |     FIL/USDT:USDT |          FIL |  USDT |   USDT |       FIL |    USDT |     USDT |   swap | false |  false |  true |  false |  false |   true |     true |   true |   false |              |        |                |        |            |                                   {} |                              {"leverage":{},"amount":{},"price":{},"cost":{}} |        
453 objects
2023-10-27T04:42:12.750Z iteration 1 passed in 3321 ms
```

```
% bigone contractPublicGetInstruments
2023-10-26T23:34:34.412Z
Node.js: v18.15.0
CCXT v4.1.29
bigone.contractPublicGetInstruments ()
2023-10-26T23:34:41.276Z iteration 0 passed in 1730 ms

 usdtPrice |  symbol | btcPrice | ethPrice | nextFundingRate | fundingRate | latestPrice | last24hPriceChange | indexPrice |   volume24h |       turnover24h | nextFundingTime |      markPrice | last24hMaxPrice | volume24hInUsd |         openValue | last24hMinPrice | openInterest
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1.00029647 |  BTCUSD | 34178.68 |  1803.91 |         0.00010 |     0.00010 |     34190.5 |              -0.01 |   34178.68 |   291994137 | 9097.532382343083 |   1698372000000 | 34179.71493043 |         34836.5 |            0.0 | 27.48238961495615 |         33752.0 |     953143.0
1.00029647 |  ETHUSD | 34178.68 |  1803.91 |         0.00010 |     0.00010 |     1802.89 |             0.0091 |    1803.91 | 6.409886E+7 |  37471.7323449977 |   1698372000000 |  1803.96462239 |          1864.3 |            0.0 |  341.016197022011 |         1762.27 |     527226.0
1.00029647 | BTCUSDT | 34178.68 |  1803.91 |         0.00010 |     0.00010 |     34171.5 |            -0.0104 |   34168.55 |    53987443 |    1741821472.769 |   1698372000000 | 34169.58462369 |         34829.5 |            0.0 |  882433.001109635 |         33754.0 |      26213.0
1.00029647 | ETHUSDT | 34178.68 |  1803.91 |         0.00010 |     0.00010 |     1803.86 |             0.0096 |    1804.56 |   6652221.0 |    1145810670.282 |   1698372000000 |  1804.61464208 |         1865.54 |            0.0 | 1166489.956689195 |         1763.76 |       7013.0
1.00029647 | EOSUSDT | 34178.68 |  1803.91 |         0.00010 |     0.00010 |         0.6 |             0.0068 |     0.5978 |   4274075.0 |       24828103.01 |   1698372000000 |      0.5978181 |           0.618 |            0.0 |        18478.2035 |           0.575 |       2508.0
1.00029647 | DOTUSDT | 34178.68 |  1803.91 |         0.00010 |     0.00010 |       4.233 |            -0.0128 |     4.2359 |    188258.0 |        772867.626 |   1698372000000 |     4.23602826 |           4.452 |            0.0 |  416011.637486965 |           4.159 |      92228.0
1.00029647 | UNIUSDT | 34178.68 |  1803.91 |         0.00010 |     0.00010 |       4.104 |            -0.0153 |     4.1031 |    128994.0 |        535582.816 |   1698372000000 |     4.10322424 |           4.376 |            0.0 |  165309.409963095 |           4.005 |      32711.0
1.00029647 | FILUSDT | 34178.68 |  1803.91 |         0.00010 |     0.00010 |       3.738 |            -0.0076 |     3.7337 |    237513.0 |        871489.892 |   1698372000000 |     3.73381306 |           3.877 |            0.0 |       27560.44495 |           3.624 |       4124.0
8 objects
2023-10-26T23:34:41.276Z iteration 1 passed in 1730 ms
```

```
% bigone contractPublicGetDepthSymbolSnapshot '{"symbol": "FILUSDT"}'
2023-10-26T23:43:06.531Z
Node.js: v18.15.0
CCXT v4.1.29
bigone.contractPublicGetDepthSymbolSnapshot ([object Object])
2023-10-26T23:43:12.483Z iteration 0 passed in 1583 ms

{
  asks: {
    '3.745': '644',
    ...
    '4.345': '10332'
  },
  bids: {
    '3.715': '11',
    ...
    '3.599': '781'
  },
  to: '61252406',
  lastPrice: '3.736',
  bestPrices: { ask: '3.744', bid: '3.724' },
  from: '0'
}
2023-10-26T23:43:12.483Z iteration 1 passed in 1583 ms
```

```
2023-10-27T02:14:34.726Z
Node.js: v18.15.0
CCXT v4.1.29
bigone.contractPrivateGetAccounts ()
2023-10-27T02:14:42.734Z iteration 0 passed in 1658 ms

           cash |                                                                                       positions
-----------------------------------------------------------------------------------------------------------------
[object Object] |                                                                                 [object Object]
[object Object] |                                                                                 [object Object]
[object Object] | [object Object],[object Object],[object Object],[object Object],[object Object],[object Object]
3 objects
2023-10-27T02:14:42.734Z iteration 1 passed in 1658 ms
```

### CODE REVIEW

File: ts/src/bigone.ts [hunk 1]
```diff
@@ -577,6 +635,65 @@ export default class bigone extends Exchange {
...
+            const market = contractResponse[i];
+            const marketId = this.safeString (market, 'symbol');
+            const baseId = marketId.slice (0, 3);
+            const quoteId = marketId.slice (3);
+            const inverse = (quoteId === 'USD');
+            const settleId = inverse ? baseId : quoteId;
+            const base = this.safeCurrencyCode (baseId);
+            const quote = this.safeCurrencyCode (quoteId);
+            const settle = this.safeCurrencyCode (settleId);
+            result.push ({
```
Comment (carlosmiei) [inline ts/src/bigone.ts:648]: @samgermain we now have a `safeMarketStructure` that we can pass the markets through

**Comment by carlosmiei** [comment]
@samgermain can you check the build pls? *If needed you can use skip-tests.json to skip some of the checks
image

**Comment by samgermain** [comment]
> @samgermain can you check the build pls? *If needed you can use skip-tests.json to skip some of the checks image

i don't get the error locally, it's also just more efficient to push it and then check if the pr build fails. If it does I'll fix the error, so you don't need to open the build logs, I'll do that.

**Comment by samgermain** [comment]
How do we solve this error `[TEST_FAILURE] Exchange can not load markets [ExchangeError] ExchangeError: bigone does not have currency code USD`, I think it's happening because USD is used in the market of a contract (e.g. BTC/USD:BTC, but isn't a currency on the exchange so it's not returned in fetchCurrencies

File: ts/src/bigone.ts [hunk 2]
```diff
@@ -1709,7 +1829,7 @@ export default class bigone extends Exchange {
         };
     }
 
-    async fetchDeposits (code: Str = undefined, since: Int = undefined, limit: Int = undefined, params = {}): Promise<Transaction[]> {
+    async fetchDeposits (code: string = undefined, since: Int = undefined, limit: Int = undefined, params = {}): Promise<any> {
```
Comment (carlosmiei) [inline ts/src/bigone.ts:1832]: @samgermain why are you removing some types from the signature, like Str and the return type?

Comment (carlosmiei) [reply ts/src/bigone.ts:1832]: @samgermain only ⬆️  comment is preventing the merge

**Pull request merged**